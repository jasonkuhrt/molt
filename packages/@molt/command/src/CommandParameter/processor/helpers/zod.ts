import type { Pam } from '../../../lib/Pam/index.js'
import type { ZodNumberCheck, ZodStringCheck } from '../../../lib/zodHelpers/index_.js'
import type { SomeBasicType, SomeUnionType, TypeNumber, TypeString } from '../../types.js'
import { Alge } from 'alge'
import { z } from 'zod'

// export const analyzeZodType = (zodType: SomeBasicType | SomeUnionType) => {
//   if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodUnion) {
//     let description = zodType.description ?? null
//     let primitiveType = zodType

//     while (
//       primitiveType._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault ||
//       primitiveType._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional
//     ) {
//       description = description ?? primitiveType._def.innerType.description ?? null
//       primitiveType = primitiveType._def.innerType
//     }

//     const members = zodType._def.options.map((_) => {
//       const typeAnalysis = analyzeZodTypeScalar(_)
//       return {
//         // zodType: _,
//         description: typeAnalysis.description,
//         type: typeAnalysis.type,
//       }
//     })
//     return {
//       _tag: `TypeUnion`,
//       members,
//     }
//   }

//   return analyzeZodTypeScalar(zodType)
// }

export const analyzeZodType = (zodType: SomeBasicType | SomeUnionType) => {
  return analyzeZodType_(zodType, zodType.description ?? null)
}

// prettier-ignore
type ZodTypeToType<ZT extends SomeBasicType | SomeUnionType> =
  ZT extends z.ZodOptional<infer T>       ? ZodTypeToType<T> :
  ZT extends z.ZodDefault<infer T>        ? ZodTypeToType<T> :
  ZT extends z.ZodLiteral<any>            ? Pam.Type.Literal :
  ZT extends z.ZodString                  ? Pam.Type.String :
  ZT extends z.ZodBoolean                 ? Pam.Type.Boolean :
  ZT extends z.ZodNumber                  ? Pam.Type.Number :
  ZT extends z.ZodEnum<any>               ? Pam.Type.Enumeration :
  ZT extends z.ZodNativeEnum<any>         ? Pam.Type.Enumeration :
  ZT extends z.ZodUnion<any>              ? Pam.Type.Union :
                                            never

const analyzeZodType_ = <ZT extends SomeBasicType | SomeUnionType>(
  zodType: ZT,
  description: null | string,
): {
  type: ZodTypeToType<ZT>
  // type: ZT
  description: null | string
} => {
  if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault) {
    return analyzeZodType_(zodType._def.innerType as ZT, description)
  }

  if (zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional) {
    return analyzeZodType_(zodType._def.innerType as ZT, description)
  }

  const type: null | Pam.Type.Group.Any =
    zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodUnion
      ? {
          _tag: `TypeUnion` as const,
          members: zodType._def.options.map((_) => {
            const typeAnalysis = analyzeZodType(_)
            return {
              // zodType: _,
              description: typeAnalysis.description,
              type: typeAnalysis.type,
            }
          }),
        }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodLiteral
      ? { _tag: `TypeLiteral`, value: zodType._def.value }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodString
      ? { _tag: `TypeString`, ...mapZodStringChecks(zodType._def.checks) }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodBoolean
      ? { _tag: `TypeBoolean` }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodNumber
      ? { _tag: `TypeNumber`, ...mapZodNumberChecks(zodType._def.checks) }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodNativeEnum
      ? { _tag: `TypeEnum`, members: Object.values(zodType._def.values) }
      : zodType._def.typeName === z.ZodFirstPartyTypeKind.ZodEnum
      ? { _tag: `TypeEnum`, members: zodType._def.values }
      : null

  if (!type) throw new Error(`Unsupported zod type: ${zodType._def.typeName}`)

  return {
    description,
    type: type as ZodTypeToType<ZT>,
  }
}

const mapZodNumberChecks = (checks: z.ZodNumberCheck[]) => {
  return checks
    .map((_): ZodNumberCheck => ({ _tag: _.kind, ..._ }) as any)
    .reduce(
      (acc, check) => {
        return Alge.match(check)
          .int(() => ({ ...acc, int: true }))
          .min((check) => ({
            min: check.value,
          }))
          .max((check) => ({
            max: check.value,
          }))
          .finite(() => ({
            finite: true,
          }))
          .multipleOf((check) => ({
            multipleOf: check.value,
          }))
          .done()
      },
      {} as Omit<TypeNumber, '_tag'>,
    )
}

const mapZodStringChecks = (checks: z.ZodStringCheck[]): Omit<TypeString, '_tag'> => {
  return checks
    .map((_): ZodStringCheck => ({ _tag: _.kind, ..._ }) as any)
    .reduce<Omit<TypeString, '_tag'>>(
      (acc, check) => {
        return {
          ...acc,
          ...Alge.match(check)
            .regex((check) => ({
              regex: check.regex,
            }))
            .min((check) => ({
              min: check.value,
            }))
            .max((check) => ({
              max: check.value,
            }))
            .url((check) => ({
              pattern: { type: check._tag },
            }))
            .cuid((check) => ({
              pattern: { type: check._tag },
            }))
            .cuid2(() => ({
              pattern: { type: `cuid2` as const },
            }))
            .uuid((check) => ({
              pattern: { type: check._tag },
            }))
            .emoji((_) => ({
              pattern: { type: _._tag },
            }))
            .ip((_) => ({
              pattern: {
                type: _._tag,
                version: _.version
                  ? Alge.match(_.version)
                      .v4(() => 4 as const)
                      .v6(() => 6 as const)
                      .done()
                  : null,
              },
            }))
            .ulid((_) => ({
              pattern: { type: _._tag },
            }))
            .datetime((check) => ({
              pattern: {
                type: `dateTime` as const,
                offset: check.offset,
                precision: check.precision,
              },
            }))
            .email((check) => ({
              pattern: { type: check._tag },
            }))
            .endsWith((check) => ({
              endsWith: check.value,
            }))
            .startsWith((check) => ({
              startsWith: check.value,
            }))
            .length((check) => ({
              length: check.value,
            }))
            .includes((_) => ({
              includes: _.value,
            }))
            // transformations
            .trim(() => ({
              transformations: {
                ...acc.transformations,
                trim: true,
              },
            }))
            .toLowerCase(() => ({
              transformations: {
                ...acc.transformations,
                toCase: `lower` as const,
              },
            }))
            .toUpperCase(() => ({
              transformations: {
                ...acc.transformations,
                toCase: `upper` as const,
              },
            }))
            .done(),
        }
      },
      {} as Omit<TypeString, '_tag'>,
    )
}
