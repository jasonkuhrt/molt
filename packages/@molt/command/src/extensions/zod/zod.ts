import type { CommandParameter } from '../../CommandParameter/index.js'
import { createExtension } from '../../extension.js'
import type { HKT } from '../../helpers.js'
import { TypeAdaptor } from './TypeAdaptor/index.js'
import type { z } from 'zod'

export type ZodType = CommandParameter.SomeBasicType | CommandParameter.SomeUnionType

interface ZodTypeMapper extends HKT.Fn<z.ZodTypeAny> {
  return: TypeAdaptor.Zod.FromZod<this['params']>
}

export const Zod = createExtension<ZodType, ZodTypeMapper>({
  name: `Zod`,
  type: (zodType: z.ZodFirstPartySchemaTypes) => TypeAdaptor.Zod.fromZod(zodType) as any,
})
