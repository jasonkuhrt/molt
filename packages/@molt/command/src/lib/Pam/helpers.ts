import type { Type } from './kinds.js'

// prettier-ignore
export type TypeToValueMapping<T extends Type> =
  T extends Type.Scalar.Boolean         ? boolean :
  T extends Type.Scalar.Number          ? number  :
  T extends Type.Scalar.String          ? string  :
  T extends Type.Scalar.Enumeration     ? string  :
  T extends Type.Union                  ? boolean | number | string :
                                          never
