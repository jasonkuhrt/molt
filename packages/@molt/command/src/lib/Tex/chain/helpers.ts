import type { Block } from '../nodes/block.js'
import type { List } from '../nodes/list.js'
import type { Node } from '../nodes/node.js'
import type { Table } from '../nodes/table.js'
import type { BlockBuilder } from './block.js'
import type { ListBuilder } from './list.js'
import type { RootBuilder } from './root.js'
import type { TableBuilder } from './table.js'

export type NodeImplementor<B extends Builder> = ($: B) => null | B

export type Builder = RootBuilder | BlockBuilder | TableBuilder | ListBuilder

export interface BuilderInternal<N = Node> {
  _: {
    node: N
  }
}

// prettier-ignore
export const toInternalBuilder = <Builder extends BlockBuilder<null>|TableBuilder|ListBuilder|null>(builder: Builder):
  Builder extends null               ? null                   :
  Builder extends BlockBuilder<null> ? BuilderInternal<Block> :
  Builder extends TableBuilder       ? BuilderInternal<Table> :
  Builder extends ListBuilder        ? BuilderInternal<List>  :
                                       never => 
  builder as any
