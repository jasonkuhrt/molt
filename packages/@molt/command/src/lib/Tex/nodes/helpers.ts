import { Text } from '../../Text/index.js'
import type { BlockParameters } from './block.js'

export interface RenderContext {
  maxWidth?: undefined | number
  height?: undefined | number
  color?: undefined | ((text: string) => string)
  phase?: undefined | 'inner' | 'outer'
  index: {
    total: number
    isLast: boolean
    isFirst: boolean
    position: number
  }
}

export interface Shape {
  intrinsicWidth: number
  intrinsicHeight: number
  desiredWidth: number | null
}

export const applyPadding = (
  value: string,
  padding: Exclude<BlockParameters['padding'], undefined>,
  context: Pick<RenderContext, 'index'>,
) => {
  value =
    padding.topBetween && !context.index.isFirst
      ? Text.chars.newline.repeat(padding.topBetween) + value
      : value
  value = padding.top ? Text.chars.newline.repeat(padding.top) + value : value
  value = padding.left ? Text.indentBlock(value, Text.chars.space.repeat(padding.left)) : value
  value = padding.bottom ? value + Text.chars.newline.repeat(padding.bottom) : value
  value =
    padding.bottomBetween && !context.index.isLast
      ? value + Text.chars.newline.repeat(padding.bottomBetween)
      : value
  value = padding.right
    ? Text.fromLines(Text.toLines(value).map((_) => _ + Text.chars.space.repeat(padding.right!)))
    : value

  return value
}
