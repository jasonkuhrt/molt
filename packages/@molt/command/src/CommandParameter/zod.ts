import { ZodHelpers } from '../lib/zodHelpers/index.js'
import { stripOptionalAndDefault } from '../lib/zodHelpers/index_.js'
import { z } from 'zod'

export type SomeBasicParameterType = SomeBasicType | SomeUnionType

type ZodEnumBase = z.ZodEnum<[string, ...string[]]>

type ZodNativeEnumBase = z.ZodNativeEnum<any>

export type SomeBasicType =
  | SomeBasicTypeScalar
  | z.ZodOptional<SomeBasicTypeScalar>
  | z.ZodDefault<SomeBasicTypeScalar>

export type SomeUnionType = SomeUnionTypeScalar | z.ZodOptional<SomeUnionType> | z.ZodDefault<SomeUnionType>

export type SomeExclusiveZodType = SomeBasicTypeScalar

export type SomeUnionTypeScalar = z.ZodUnion<
  [SomeBasicTypeScalar, SomeBasicTypeScalar, ...SomeBasicTypeScalar[]]
>

// prettier-ignore
export type SomeBasicTypeScalar =
  | z.ZodString
  | ZodEnumBase
  | ZodNativeEnumBase
  | z.ZodNumber
  | z.ZodBoolean
  | z.ZodLiteral<number>
  | z.ZodLiteral<string>
  | z.ZodLiteral<boolean>

export const isUnionType = (type: SomeBasicType | SomeUnionType): type is SomeUnionType => {
  const type_ = stripOptionalAndDefault(type)
  const isUnion = type_._def.typeName === z.ZodFirstPartyTypeKind.ZodUnion
  return isUnion
}

export const getUnionScalar = (zodType: SomeUnionType): SomeUnionTypeScalar => {
  const zodScalarType = ZodHelpers.stripOptionalAndDefault(zodType)
  return zodScalarType
}

export const getBasicScalar = (zodType: SomeBasicType): SomeBasicTypeScalar => {
  const type = ZodHelpers.stripOptionalAndDefault(zodType)
  return type
}
