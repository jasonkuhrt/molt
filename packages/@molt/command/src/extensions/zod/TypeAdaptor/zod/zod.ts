import { Type } from '../../../../Type/index.js'
import { ZodHelpers } from '../../zodHelpers/index.js'
import { Alge } from 'alge'
import type { Simplify } from 'type-fest'
import { z } from 'zod'

// prettier-ignore
export type FromZod<ZodType extends z.ZodType> =
  ZodType extends z.ZodOptional<infer T>
    ? Type.Union<[FromZodNonOptional<T>, Type.Literal<undefined>]>
    : FromZodNonOptional<ZodType>

// prettier-ignore
export type FromZodNonOptional<ZodType extends z.ZodType> =
  ZodType extends ZodTypeScalar              ? FromZodScalar<ZodType> :
  ZodType extends z.ZodDefault<infer T>      ? FromZodScalar<T> :
  ZodType extends z.ZodUnion<infer T>        ? Type.Union<{[i in keyof T]: FromZodScalar<T[i]>}> :
                                                never

// prettier-ignore
export type FromZodScalar<ZodType extends ZodTypeScalar> =
  ZodType extends z.ZodString                                           ? Type.Scalar.String :
  ZodType extends z.ZodBoolean                                          ? Type.Scalar.Boolean :
  ZodType extends z.ZodLiteral<infer T extends boolean|string|number>   ? Type.Scalar.Literal<T> :
  ZodType extends z.ZodNumber                                           ? Type.Scalar.Number :
  ZodType extends z.ZodEnum<infer T>                                    ? Type.Scalar.Enumeration<T> :
  ZodType extends z.ZodNativeEnum<infer T extends z.EnumLike>           ? Type.Scalar.Enumeration<EnumerationMembersFromZodEnumLike<T>> :
                                                                          never
type EnumerationMembersFromZodEnumLike<T extends z.EnumLike> = T[keyof T][]

type ZodTypeScalar =
  | z.ZodString
  | z.ZodBoolean
  | z.ZodLiteral<any>
  | z.ZodNumber
  | z.ZodEnum<any>
  | z.ZodNativeEnum<any>
// prettier-ignore

export const fromZod = (zodType: z.ZodFirstPartySchemaTypes): Type.Type => _fromZod(zodType, {})

// prettier-ignore
const _fromZod = (zodType: z.ZodFirstPartySchemaTypes,previous:{description?:string;optionality?:Type.Optionality<any>}): Type.Type => {
  const zt = zodType
  const description = previous.description ?? zt.description

  const isOptional = zt._def.typeName === z.ZodFirstPartyTypeKind.ZodOptional
  const hasDefault = zt._def.typeName === z.ZodFirstPartyTypeKind.ZodDefault
  const defaultGetter = hasDefault ? (zt._def.defaultValue as () => unknown) : null
  const optionality: Type.Optionality<any> = previous.optionality ?? (defaultGetter
    ? { _tag: `default`, getValue: () => defaultGetter() }
    : isOptional
    ? { _tag: `optional` }
    : { _tag: `required` })
  
  if (ZodHelpers.isString(zt)) {
    const {refinements,transformations} = mapZodStringChecksAndTransformations(zt._def.checks)
    return Type.string({ optionality, refinements, transformations, description })
  }
  if (ZodHelpers.isNumber(zt)) {
    const {refinements} = mapZodNumberChecksAndTransformations(zt._def.checks)
    return Type.number({ optionality, refinements, description })
  }
  if (ZodHelpers.isEnum(zt))            return Type.enumeration({ optionality, members: zt._def.values, description }) // eslint-disable-line
  if (ZodHelpers.isNativeEnum(zt))      return Type.enumeration({ optionality, members: Object.values<any>(zt._def.values), description })
  if (ZodHelpers.isBoolean(zt))         return Type.boolean({ optionality, description })
  if (ZodHelpers.isLiteral(zt))         return Type.literal({ optionality, value: zt._def.value, description }) // eslint-disable-line
  if (ZodHelpers.isDefault(zt))         return _fromZod(zt._def.innerType,{...previous,description,optionality})
  if (ZodHelpers.isOptional(zt))        return _fromZod(zt._def.innerType,{...previous,description,optionality})
  if (ZodHelpers.isUnion(zt))           {
    if (!Array.isArray(zt._def.options)) throw new Error(`Unsupported zodType: ${JSON.stringify(zt[`_def`])}`)
    const members = zt._def.options.map((_: {_def:{description:undefined|string}}) => {
      const description = _._def.description
      return _fromZod(_ as any, {...previous,description,optionality})
    })
    return Type.union({ optionality, members_: members, description })
  }
  throw new Error(`Unsupported zodType: ${JSON.stringify(zt[`_def`])}`)
}

const mapZodNumberChecksAndTransformations = (checks: z.ZodNumberCheck[]) => {
  return checks.reduce(
    (acc, check) => {
      return {
        refinements: {
          ...acc.refinements,
          ...Alge.match(check)
            .int(() => ({ ...acc, int: true }))
            .min((check) => ({
              min: check.value,
            }))
            .max((check) => ({
              max: check.value,
            }))
            .finite(() => ({
              finite: true,
            }))
            .multipleOf((check) => ({
              multipleOf: check.value,
            }))
            .done(),
        },
      }
    },
    {} as Simplify<Pick<Type.Scalar.Number, 'refinements'>>,
  )
}

const mapZodStringChecksAndTransformations = (checks: z.ZodStringCheck[]) => {
  const transformations = [`trim`, `toCase`] as const
  return checks.reduce(
    (acc, check) => {
      const refinementOrTransformation = Alge.match(check)
        .regex((check) => ({
          regex: check.regex,
        }))
        .min((check) => ({
          min: check.value,
        }))
        .max((check) => ({
          max: check.value,
        }))
        .url((check) => ({
          pattern: { type: check.kind },
        }))
        .cuid((check) => ({
          pattern: { type: check.kind },
        }))
        .cuid2(() => ({
          pattern: { type: `cuid2` as const },
        }))
        .uuid((check) => ({
          pattern: { type: check.kind },
        }))
        .emoji((_) => ({
          pattern: { type: _.kind },
        }))
        .ip((_) => ({
          pattern: {
            type: _.kind,
            version: _.version
              ? Alge.match(_.version)
                  .v4(() => 4 as const)
                  .v6(() => 6 as const)
                  .done()
              : null,
          },
        }))
        .ulid((_) => ({
          pattern: { type: _.kind },
        }))
        .datetime((check) => ({
          pattern: {
            type: `dateTime` as const,
            offset: check.offset,
            precision: check.precision,
          },
        }))
        .email((check) => ({
          pattern: { type: check.kind },
        }))
        .endsWith((check) => ({
          endsWith: check.value,
        }))
        .startsWith((check) => ({
          startsWith: check.value,
        }))
        .length((check) => ({
          length: check.value,
        }))
        .includes((_) => ({
          includes: _.value,
        }))
        // transformations
        .trim(() => ({
          trim: true,
        }))
        .toLowerCase(() => ({
          toCase: `lower` as const,
        }))
        .toUpperCase(() => ({
          toCase: `upper` as const,
        }))
        .done()

      const isTransformation = transformations.includes(Object.keys(refinementOrTransformation)[0]! as any)
      return isTransformation
        ? {
            ...acc,
            transformations: {
              ...acc.transformations,
              ...refinementOrTransformation,
            },
          }
        : {
            ...acc,
            refinements: {
              ...acc.refinements,
              ...refinementOrTransformation,
            },
          }
    },
    {} as Simplify<Pick<Type.Scalar.String, 'refinements' | 'transformations'>>,
  )
}
