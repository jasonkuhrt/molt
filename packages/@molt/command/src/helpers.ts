import camelCase from 'lodash.camelcase'
import { z } from 'zod'

export const BooleanLookup = {
  true: true,
  false: false,
} as const

export const environmentVariableBooleanLookup = {
  ...BooleanLookup,
  '1': true,
  '0': false,
} as const

export const stripeDashPrefix = (flagNameInput: string): string => {
  return flagNameInput.replace(/^-+/, ``)
}

export const zodPassthrough = <T>() => z.any().transform((_) => _ as T)

export type Values<T> = T[keyof T]

export const getLowerCaseEnvironment = (): NodeJS.ProcessEnv => lowerCaseObjectKeys(process.env)

export const lowerCaseObjectKeys = (obj: object) =>
  Object.fromEntries(Object.entries(obj).map(([k, v]) => [k.toLowerCase(), v]))

export const parseEnvironmentVariableBoolean = (value: string): boolean | null =>
  // @ts-expect-error ignore
  // eslint-disable-next-line
  environmentVariableBooleanLookup[value] ?? null

export const parseEnvironmentVariableBooleanOrThrow = (value: string) => {
  const result = parseEnvironmentVariableBoolean(value)
  if (result === null) {
    throw new Error(`Invalid boolean value: ${value}`)
  }
  return result
}

export const negateNamePattern = /^no([A-Z].+)/

export const stripeNegatePrefix = (name: string): null | string => {
  // eslint-disable-next-line
  const withoutPrefix = name.match(negateNamePattern)?.[1]!
  if (!withoutPrefix) return null
  const withCamelCase = camelCase(withoutPrefix)
  return withCamelCase
}

export const stripeNegatePrefixLoose = (name: string): string => {
  const result = stripeNegatePrefix(name)
  return result ? result : name
}

export const invertTable = <T>(rows: T[][]): T[][] => {
  const columns: T[][] = []
  for (const row of rows) {
    let i = 0
    for (const col of row) {
      const column = columns[i] || []
      column.push(col)
      columns[i] = column
      i++
    }
  }
  return columns
}

export const entries = <O extends Record<string, unknown>>(
  obj: O,
): Exclude<{ [k in keyof O]: [k, O[k]] }[keyof O], undefined>[] => Object.entries(obj) as any

export const casesExhausted = (_: never): never => {
  throw new Error(`Cases exhausted: ${_}`) // eslint-disable-line
}
