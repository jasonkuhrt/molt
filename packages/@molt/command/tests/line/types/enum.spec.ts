import { describe, expect, it } from 'vitest'
import { z } from 'zod'
import { $, s } from '../../_/helpers.js'
import { stdout } from '../../_/mocks.js'

describe(`errors`, () => {
  it(`when argument missing (last position)`, () => {
    $.parameter(`--mode`, z.enum([`a`, `b`])).parse({ line: [`--mode`] })
    expect(stdout.mock.calls).toMatchSnapshot()
  })
  it(`when argument missing (non-last position)`, () => {
    // prettier-ignore
    $.parameter(`--name`, s).parameter(`--mode`, z.enum([`a`, `b`])).parse({ line: [` --mode`, `--name`, `joe`] })
    expect(stdout.mock.calls).toMatchSnapshot()
  })
  it(`is validated`, () => {
    // const args = Parameters.create({ '--mode': z.enum([`a`, `b`, `c`]) }).parse({line:[`--mode`, `bad`]})
    // assert<IsExact<{ mode: 'a'|'b'|'c' }, typeof args>>(true)
    // expect(args).toMatchObject({ mode: true })
    $.parameter(`--mode`, z.enum([`a`, `b`, `c`])).parse({ line: [`--mode`, `bad`] })
    expect(stdout.mock.calls).toMatchSnapshot()
  })
})
