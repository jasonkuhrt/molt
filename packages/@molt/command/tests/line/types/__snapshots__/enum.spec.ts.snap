// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`errors > is validated 1`] = `
[
  [
    "Cannot run command, you made some mistakes:

Invalid value for mode:
Value is not a member of the enum.

Here are the docs for this command:

",
  ],
  [
    "
PARAMETERS
                                                                      
  [4m[90mName[39m[24m     [4m[90mType[39m[24m                 [4m[90mDefault[39m[24m       [4m[90mEnvironment (1)[39m[24m
                                                             
  [1m[32mmode[39m[22m     [32ma[39m[33m | [39m[32mb[39m[33m | [39m[32mc[39m            [1m[30m[101m REQUIRED [49m[39m[22m    [34m✓[39m              
                                                             
  [2m[90m[39m[22m
  [2m[90mNOTES                                                                           [39m[22m
  [2m[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━[39m[22m
  [2m[90m(1) Parameters can be passed arguments via environment variables. Command line  [39m[22m
  [2m[90m    arguments take precedence. Environment variable names are snake cased       [39m[22m
  [2m[90m    versions of the parameter name (or its aliases), case insensitive. They must[39m[22m
  [2m[90m    be prefixed with [34mCLI_PARAMETER_[39m[90m or [34mCLI_PARAM_[39m[90m (case insensitive). Examples: [39m[22m
  [2m[90m      → [34mCLI_PARAMETER_[39m[90m[32mMODE[39m[90m=\\"...\\"                                                [39m[22m
",
  ],
]
`;

exports[`errors > when argument missing (last position) 1`] = `
[
  [
    "Cannot run command, you made some mistakes:

Missing argument

Here are the docs for this command:

",
  ],
  [
    "
PARAMETERS
                                                                      
  [4m[90mName[39m[24m     [4m[90mType[39m[24m             [4m[90mDefault[39m[24m       [4m[90mEnvironment (1)[39m[24m
                                                         
  [1m[32mmode[39m[22m     [32ma[39m[33m | [39m[32mb[39m            [1m[30m[101m REQUIRED [49m[39m[22m    [34m✓[39m              
                                                         
  [2m[90m[39m[22m
  [2m[90mNOTES                                                                           [39m[22m
  [2m[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━[39m[22m
  [2m[90m(1) Parameters can be passed arguments via environment variables. Command line  [39m[22m
  [2m[90m    arguments take precedence. Environment variable names are snake cased       [39m[22m
  [2m[90m    versions of the parameter name (or its aliases), case insensitive. They must[39m[22m
  [2m[90m    be prefixed with [34mCLI_PARAMETER_[39m[90m or [34mCLI_PARAM_[39m[90m (case insensitive). Examples: [39m[22m
  [2m[90m      → [34mCLI_PARAMETER_[39m[90m[32mMODE[39m[90m=\\"...\\"                                                [39m[22m
",
  ],
]
`;

exports[`errors > when argument missing (non-last position) 1`] = `
[
  [
    "Cannot run command, you made some mistakes:

Missing argument

Here are the docs for this command:

",
  ],
  [
    "
PARAMETERS
                                                                      
  [4m[90mName[39m[24m     [4m[90mType[39m[24m              [4m[90mDefault[39m[24m       [4m[90mEnvironment (1)[39m[24m
                                                          
  [1m[32mmode[39m[22m     [32ma[39m[33m | [39m[32mb[39m             [1m[30m[101m REQUIRED [49m[39m[22m    [34m✓[39m              
                                                          
  [1m[32mname[39m[22m     [32mstring[39m            [1m[30m[101m REQUIRED [49m[39m[22m    [34m✓[39m              
                                                          
  [2m[90m[39m[22m
  [2m[90mNOTES                                                                           [39m[22m
  [2m[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━[39m[22m
  [2m[90m(1) Parameters can be passed arguments via environment variables. Command line  [39m[22m
  [2m[90m    arguments take precedence. Environment variable names are snake cased       [39m[22m
  [2m[90m    versions of the parameter name (or its aliases), case insensitive. They must[39m[22m
  [2m[90m    be prefixed with [34mCLI_PARAMETER_[39m[90m or [34mCLI_PARAM_[39m[90m (case insensitive). Examples: [39m[22m
  [2m[90m      → [34mCLI_PARAMETER_[39m[90m[32mMODE[39m[90m=\\"...\\"                                                [39m[22m
  [2m[90m      → [34mCLI_PARAMETER_[39m[90m[32mNAME[39m[90m=\\"...\\"                                                [39m[22m
",
  ],
]
`;
